sdk/nodejs/x/automation/localWorkspace.ts

```diff
@@ -0,0 +1,584 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as fs from "fs";
+import * as yaml from "js-yaml";
+import * as os from "os";
+import * as upath from "upath";
+
+import { CommandResult, runPulumiCmd } from "./cmd";
+import { ConfigMap, ConfigValue } from "./config";
+import { ProjectSettings } from "./projectSettings";
+import { Stack } from "./stack";
+import { StackSettings } from "./stackSettings";
+import { PulumiFn, StackSummary, WhoAmIResult, Workspace } from "./workspace";
+
+/**
+ * LocalWorkspace is a default implementation of the Workspace interface.
+ * A Workspace is the execution context containing a single Pulumi project, a program,
+ * and multiple stacks. Workspaces are used to manage the execution environment,
+ * providing various utilities such as plugin installation, environment configuration
+ * ($PULUMI_HOME), and creation, deletion, and listing of Stacks.
+ * LocalWorkspace relies on Pulumi.yaml and Pulumi.<stack>.yaml as the intermediate format
+ * for Project and Stack settings. Modifying ProjectSettings will
+ * alter the Workspace Pulumi.yaml file, and setting config on a Stack will modify the Pulumi.<stack>.yaml file.
+ * This is identical to the behavior of Pulumi CLI driven workspaces.
+ *
+ * @alpha
+ */
+export class LocalWorkspace implements Workspace {
+    /**
+     * The working directory to run Pulumi CLI commands
+     */
+    readonly workDir: string;
+    /**
+     * The directory override for CLI metadata if set.
+     * This customizes the location of $PULUMI_HOME where metadata is stored and plugins are installed.
+     */
+    readonly pulumiHome?: string;
+    /**
+     * The secrets provider to use for encryption and decryption of stack secrets.
+     * See: https://www.pulumi.com/docs/intro/concepts/config/#available-encryption-providers
+     */
+    readonly secretsProvider?: string;
+    /**
+     *  The inline program `PulumiFn` to be used for Preview/Update operations if any.
+     *  If none is specified, the stack will refer to ProjectSettings for this information.
+     */
+    public program?: PulumiFn;
+    /**
+     * Environment values scoped to the current workspace. These will be supplied to every Pulumi command.
+     */
+    public envVars: { [key: string]: string };
+    private ready: Promise<any[]>;
+    /**
+     * Creates a workspace using the specified options. Used for maximal control and customization
+     * of the underlying environment before any stacks are created or selected.
+     *
+     * @param opts Options used to configure the Workspace
+     */
+    public static async create(opts: LocalWorkspaceOptions): Promise<LocalWorkspace> {
```

Comment:
The one thing I keep wondering about is whether we can make it more intuitive/obvious which methods should be used for which scenarios.

Right now it's not immediately obvious when I should use `LocalWorkspace.create` vs `LocalWorkspace.createStack/createOrSelectStack/selectStack` vs. `Stack.create/createOrSelect/select`, without looking at example usage or reading the doc comments. If `LocalWorkspace.createStack/createOrSelectStack/selectStack` are the mainline methods to use, wondering if we should rename `create` to a name that's a little less enticing. Also when I see `LocalWorkspace.create` I wonder whether it's creating the workspace on disk or just opening an existing directory.

I'm fine shipping an initial alpha version of the APIs as they are now, but would like to noodle on this some more to see if we can come up with more intuitive entry points into this API.

Reply:
I do think that the answer here is documentation. 

Happy to revisit this later if we come up with better ideas.